Shopify/cli
# Add timeout and retry logic to downloading Javy and function-runner

### WHY are these changes introduced?

CI is sometimes marking tests around this logic as timing out. I am making an educated guess that it's the call to `fetch` that's taking more than 13 seconds to return because I don't know why ungzipping or running a subprocess's help command (the other two I/O operations) would take more than 13 seconds to complete. Adding an explicit timeout and retries to the downloading operation mean, if the downloads were to hang for whatever reason, we'd abort the download and retry 3 times. Hopefully this should be sufficient to work around this problem (assuming it is a result of the network request that's causing the timeout).

### WHAT is this pull request doing?

When downloading the Javy or function-runner binaries, only allow up to 5 seconds for the download to complete and then retry for a given number of times.

I'm also having the CLI download Javy and function-runner to a temporary directory first and then move the file when it's finished being processed so retrying the command won't keep trying to load the same broken binary. This should also help if multiple things are trying to download Javy or function-runner in parallel.

I'm open to suggestions on if that 5 second timeout I'm imposing should be higher to not make life difficult for someone who has a really bad network connection but needs to download and install Javy or function-runner at that point in time. We can get at least one retry in if the timeout is under 12.5 seconds.

### How to test your changes?

1. `rm packages/app/src/cli/services/bin/javy` (if it exists)
2. `rm packages/app/dist/cli/services/bin/javy ` (if it exists)
3. `pnpm shopify app build --path <path_to_app_with_shopify_function> --verbose`
4. There should've been a verbose log saying it was downloading Javy
5. `packages/app/dist/cli/services/bin/javy` should now exist

Repeat the above steps for running a function to test function-runner.

### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
